generate/authorize.js

```diff
@@ -0,0 +1,834 @@
+      return LIST_OF_STRINGS;
+    
+    if (field.type.type.type.name.value === USER)
+      return LIST_OF_USERS;
+
+  }
+
+  // pattern: 'coauthors: [User!]'
+  if (field.type && 
+      field.type.kind &&
+      field.type.kind === LIST_TYPE &&
+      field.type.type &&
+      field.type.type.kind &&
+      field.type.type.kind === NON_NULL_TYPE &&
+      field.type.type.type &&
+      field.type.type.type.kind &&
+      field.type.type.type.kind === NAMED_TYPE &&
+      field.type.type.type.name &&
+      field.type.type.type.name.kind &&
+      field.type.type.type.name.kind === NAME &&
+      field.type.type.type.name.value){
+
+    if (field.type.type.type.name.value === STRING)
+      return LIST_OF_STRINGS;
+
+    if (field.type.type.type.name.value === USER)
+      return LIST_OF_USERS;
+  }
+
+  // pattern: 'coauthors: [User!]!'
+  if (field.type && 
+      field.type.kind &&
+      field.type.kind === NON_NULL_TYPE &&
+      field.type.type &&
+      field.type.type.kind &&
+      field.type.type.kind === LIST_TYPE &&
+      field.type.type.type &&
+      field.type.type.type.kind &&
+      field.type.type.type.kind === NON_NULL_TYPE &&
+      field.type.type.type.type &&
+      field.type.type.type.type.kind &&
+      field.type.type.type.type.kind === NAMED_TYPE &&
+      field.type.type.type.type.name &&
+      field.type.type.type.type.name.kind &&
+      field.type.type.type.type.name.kind === NAME &&
+      field.type.type.type.type.name.value){
+
+    if (field.type.type.type.type.name.value === STRING)
+      return LIST_OF_STRINGS;
+
+    if (field.type.type.type.type.name.value === USER)
+      return LIST_OF_USERS;
+
+  }
+
+  return null;
+}
+
+/*
+ * prepare roles for code generator
+ * convert array to String value
+ * replace " by '
+ * @param {array} role
+ * @return {string} roleString
+ */
+function prep ( role ) {
+  return JSON.stringify(role).replace(/\"/g, "'");
+}
+
+/*
+ * generate authorization code for mode readOne
+ * @param {boolean} authorize
+ * @param {string} typeName
+ * @param {string} userRoles
+ * @param {string} docRoles
+ * @return {string} generatedCode
+ */
+function generateAuthCodeModeReadOne ( authorize = false, typeName = '', userRoles = [], docRoles = [] ){
+  // default code
+  let generatedCode = `const { me } = context;
+    that.authorizedLoader = new DataLoader(ids => findByIds(this.collection, ids))`;
+
+  // with @authorize directive
+  if (authorize){
+
+    if (typeName === 'user') {
+      // User has to come from current class context
+      generatedCode = `const { me } = context;
+      const authQuery = queryForRoles(me, ${prep(userRoles)}, ${prep(docRoles)}, { User }, authlog('${typeName} findOneById', 'readOne', me));
+      that.authorizedLoader = new DataLoader(ids => findByIds(this.collection, ids, authQuery));`;
+    } else {
+      // User has to come from this.context.User
+      generatedCode = `const { me, User } = context;
```

Comment:
Did you try embedding the code in some templates?

I know it's a bit clunky but it seems like it's easier to modify later if it's not buried inside a big file like this.

Also, it's a bit inconsistent to do it differently here.

I'd be cool with a different way of doing templating (maybe with proper string interpolation, or using handlebars or something?) if you wanted though.

Reply:
added handlebars templating for models and resolvers, allows also to treat the user type differently then the other types. so the additional coding with bcrypt and password don't have to be entered manually any more.